src/backends/policy_engine/souffle/datalog_lowering_visitor_test.cc

```diff
@@ -0,0 +1,243 @@
+    ir::datalog::Operation(
+        ir::datalog::Symbol("UNKNOWN"), ir::datalog::Symbol("sql.MergeOp"),
+        ir::datalog::Symbol("%0.out"),
+        ir::datalog::OperandList(
+            ir::datalog::Symbol("<<ANY>>"),
+            ir::datalog::OperandList(ir::datalog::Symbol("<<ANY>>"),
+                                     ir::datalog::OperandList())),
+        ir::datalog::AttributeList()));
+
+static const IrAndDatalogOperationPairs kIrAndDatalogOperations[] = {
+    {.ir_operation = &kStringLiteralIrOperation,
+     .datalog_is_operation = &kStringLiteralDatalogIsOperation},
+    {.ir_operation = &kMergeOpIrOperation,
+     .datalog_is_operation = &kMergeOpIsOperation}};
+
+INSTANTIATE_TEST_SUITE_P(IrOperationLoweringTest, IrOperationLoweringTest,
+                         ValuesIn(kIrAndDatalogOperations));
+
+struct ModuleAndExpectedRakshaDatalogFactBuilders {
+  std::function<ir::Module()> module_fn;
+  std::function<RakshaDatalogFacts()> facts_fn;
+};
+
+class DatalogLoweringVisitModuleTest
+    : public TestWithParam<ModuleAndExpectedRakshaDatalogFactBuilders> {};
+
+TEST_P(DatalogLoweringVisitModuleTest, DatalogLoweringVisitModuleTest) {
+  const auto [module_fn, facts_fn] = GetParam();
+  ir::Module module = module_fn();
+  RakshaDatalogFacts raksha_datalog_facts = facts_fn();
+  ir::SsaNames ssa_names;
+  DatalogLoweringVisitor visitor(ssa_names);
+  module.Accept(visitor);
+  EXPECT_EQ(visitor.datalog_facts().ToDatalogString(),
+            raksha_datalog_facts.ToDatalogString());
+}
+
+static const ModuleAndExpectedRakshaDatalogFactBuilders
+    kModuleAndExpectedRakshaDatalogFacts[] = {
+        // Empty module test.
+        {.module_fn = []() { return ir::Module(); },
+         .facts_fn = []() { return RakshaDatalogFacts(); }},
+        // Test containing a single passthrough operation.
+        {.module_fn =
+             []() {
+               ir::BlockBuilder block_builder;
+               block_builder.AddOperation(kMergeOpOperator, {},
+                                          {ir::Value(ir::value::Any())});
+               ir::Module module;
+               module.AddBlock(block_builder.build());
+               return module;
+             },
+         .facts_fn =
+             []() {
+               RakshaDatalogFacts facts;
+               facts.AddIsOperationFact(
+                   ir::datalog::IsOperationFact(ir::datalog::Operation(
+                       ir::datalog::Symbol("UNKNOWN"),
+                       ir::datalog::Symbol("sql.MergeOp"),
+                       ir::datalog::Symbol("%0.out"),
+                       ir::datalog::OperandList(ir::datalog::Symbol("<<ANY>>"),
+                                                ir::datalog::OperandList()),
+                       ir::datalog::AttributeList())));
+               return facts;
+             }},
+        // Show two passthroughs chained together.
+        {.module_fn =
+             []() {
```

Comment:
Instead of lambdas, can we just have a list of `std::vector<Operation>` and build the module in the body of the test?

Reply:
I think my initial plan was to put a test exercising `BlockArgument`s, but I realize now I never got around to writing that test.

Reply:
Using vector is more readable for the test?

Reply:
I can't use a `std::vector<Operation>` because `Operation` has a `std::unique_ptr` field in `impl_module_`, which prevents it from being copyable and thus prevents it from being used directly as test data parameters.

Reply:
In that case create a some sort of helper to avoid repeating block builder setup ?

Reply:
Added a helper to reduce boilerplate.